File: packages/guards/src/secret-pattern.ts
import type {
  Guard,
  GuardResult,
  GuardContext,
} from 'open-guardrail-core';

type SecretType =
  | 'env-var'
  | 'connection-string'
  | 'private-key'
  | 'bearer-token'
  | 'basic-auth'
  | 'webhook-url'
  | 'database-url';

interface SecretPatternOptions {
  action: 'block' | 'warn' | 'mask';
  types?: SecretType[];
}

const PATTERNS: Record<SecretType, RegExp> = {
  'env-var': /(?:API_KEY|SECRET|PASSWORD|TOKEN|PRIVATE_KEY|ACCESS_KEY|AUTH)\s*[=:]\s*['"]?[\w\-./+]{8,}['"]?/gi,
  'connection-string': /(?:mongodb|mysql|postgres|postgresql|redis|amqp|mssql):\/\/[^\s'"]+/gi,
  'private-key': /-----BEGIN\s+(?:RSA|DSA|EC|OPENSSH)?\s*PRIVATE\s+KEY-----/gi,
  'bearer-token': /Bearer\s+[A-Za-z0-9\-._~+/]+=*/g,
  'basic-auth': /Basic\s+[A-Za-z0-9+/]{20,}={0,2}/g,
  'webhook-url': /https:\/\/hooks\.slack\.com\/[^\s'"]+/gi,
  'database-url': /DATABASE_URL\s*[=:]\s*['"]?[^\s'"]+['"]?/gi,
};

const MASK_LABELS: Record<SecretType, string> = {
  'env-var': '[ENV_SECRET]',
  'connection-string': '[CONNECTION_STRING]',
  'private-key': '[PRIVATE_KEY]',
  'bearer-token': '[BEARER_TOKEN]',
  'basic-auth': '[BASIC_AUTH]',
  'webhook-url': '[WEBHOOK_URL]',
  'database-url': '[DATABASE_URL]',
};

interface SecretMatch {
  type: SecretType;
  value: string;
  start: number;
  end: number;
}

const ALL_TYPES: SecretType[] = [
  'env-var', 'connection-string', 'private-key',
  'bearer-token', 'basic-auth', 'webhook-url', 'database-url',
];

function detect(text: string, types: SecretType[]): SecretMatch[] {
  const matches: SecretMatch[] = [];
  for (const type of types) {
    const re = new RegExp(PATTERNS[type].source, PATTERNS[type].flags);
    let m: RegExpExecArray | null;
    while ((m = re.exec(text)) !== null) {
      matches.push({
        type,
        value: m[0],
        start: m.index,
        end: m.index + m[0].length,
      });
    }
  }
  return matches.sort((a, b) => b.start - a.start);
}

function maskText(text: string, matches: SecretMatch[]): string {
  let result = text;
  for (const m of matches) {
    result =
      result.slice(0, m.start) +
      MASK_LABELS[m.type] +
      result.slice(m.end);
  }
  return result;
}

export function secretPattern(options: SecretPatternOptions): Guard {
  const types = options.types ?? ALL_TYPES;

  return {
    name: 'secret-pattern',
    version: '0.1.0',
    description: 'Secret and credential pattern detection',
    category: 'security',
    supportedStages: ['input', 'output'],
    async check(
      text: string,
      _ctx: GuardContext,
    ): Promise<GuardResult> {
      const start = performance.now();
      const matches = detect(text, types);
      const triggered = matches.length > 0;

      if (!triggered) {
        return {
          guardName: 'secret-pattern',
          passed: true,
          action: 'allow',
          latencyMs: Math.round(performance.now() - start),
        };
      }

      if (options.action === 'mask') {
        return {
          guardName: 'secret-pattern',
          passed: true,
          action: 'override',
          overrideText: maskText(text, matches),
          latencyMs: Math.round(performance.now() - start),
          details: {
            detected: matches.map(({ type }) => ({ type })),
          },
        };
      }

      return {
        guardName: 'secret-pattern',
        passed: false,
        action: options.action,
        message: `Secret detected: ${matches.map((m) => m.type).join(', ')}`,
        latencyMs: Math.round(performance.now() - start),
        details: {
          detected: matches.map(({ type }) => ({ type })),
          reason: 'Text contains credentials, API keys, or connection strings that should not be exposed',
        },
      };
    },
  };
}

File: python/open_guardrail/guards/secret_pattern.py
"""Secret and credential pattern detection guard."""

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_PATTERNS: dict[str, re.Pattern[str]] = {
    "env-var": re.compile(
        r"(?:API_KEY|SECRET|PASSWORD|TOKEN"
        r"|PRIVATE_KEY|ACCESS_KEY|AUTH)"
        r"\s*[=:]\s*['\"]?[\w\-./+]{8,}['\"]?",
        re.IGNORECASE,
    ),
    "connection-string": re.compile(
        r"(?:mongodb|mysql|postgres|postgresql"
        r"|redis|amqp|mssql)://[^\s'\"]+",
        re.IGNORECASE,
    ),
    "private-key": re.compile(
        r"-----BEGIN\s+(?:RSA|DSA|EC|OPENSSH)?"
        r"\s*PRIVATE\s+KEY-----",
        re.IGNORECASE,
    ),
    "bearer-token": re.compile(
        r"Bearer\s+[A-Za-z0-9\-._~+/]+=*"
    ),
    "basic-auth": re.compile(
        r"Basic\s+[A-Za-z0-9+/]{20,}={0,2}"
    ),
    "webhook-url": re.compile(
        r"https://hooks\.slack\.com/[^\s'\"]+",
        re.IGNORECASE,
    ),
    "database-url": re.compile(
        r"DATABASE_URL\s*[=:]\s*['\"]?[^\s'\"]+['\"]?",
        re.IGNORECASE,
    ),
}

_MASK_LABELS: dict[str, str] = {
    "env-var": "[ENV_SECRET]",
    "connection-string": "[CONNECTION_STRING]",
    "private-key": "[PRIVATE_KEY]",
    "bearer-token": "[BEARER_TOKEN]",
    "basic-auth": "[BASIC_AUTH]",
    "webhook-url": "[WEBHOOK_URL]",
    "database-url": "[DATABASE_URL]",
}

_ALL_TYPES = list(_PATTERNS.keys())


def _detect(
    text: str, types: list[str]
) -> list[dict]:
    matches: list[dict] = []
    for t in types:
        pat = _PATTERNS.get(t)
        if not pat:
            continue
        for m in pat.finditer(text):
            matches.append(
                {
                    "type": t,
                    "value": m.group(),
                    "start": m.start(),
                    "end": m.end(),
                }
            )
    matches.sort(key=lambda x: x["start"], reverse=True)
    return matches


def _mask_text(text: str, matches: list[dict]) -> str:
    result = text
    for m in matches:
        label = _MASK_LABELS.get(m["type"], "[SECRET]")
        result = (
            result[: m["start"]] + label + result[m["end"] :]
        )
    return result


class _SecretPattern:
    def __init__(
        self,
        *,
        action: str = "block",
        types: Optional[List[str]] = None,
    ) -> None:
        self.name = "secret-pattern"
        self.action = action
        self._types = types or list(_ALL_TYPES)

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        matches = _detect(text, self._types)
        elapsed = (time.perf_counter() - start) * 1000

        if not matches:
            return GuardResult(
                guard_name="secret-pattern",
                passed=True,
                action="allow",
                latency_ms=round(elapsed, 2),
            )

        detected = [{"type": m["type"]} for m in matches]

        if self.action == "mask":
            return GuardResult(
                guard_name="secret-pattern",
                passed=True,
                action="override",
                override_text=_mask_text(text, matches),
                latency_ms=round(elapsed, 2),
                details={"detected": detected},
            )

        types_found = ", ".join(
            m["type"] for m in matches
        )
        return GuardResult(
            guard_name="secret-pattern",
            passed=False,
            action=self.action,
            message=f"Secret detected: {types_found}",
            latency_ms=round(elapsed, 2),
            details={
                "detected": detected,
                "reason": (
                    "Text contains credentials, API"
                    " keys, or connection strings"
                    " that should not be exposed"
                ),
            },
        )


def secret_pattern(
    *,
    action: str = "block",
    types: Optional[List[str]] = None,
) -> _SecretPattern:
    return _SecretPattern(action=action, types=types)
